text/0000-macros.md

```diff
@@ -0,0 +1,141 @@
+- Feature Name: macro
+- Start Date: 2016-04-17
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Macros by example 2.0. A replacement for `macro_rules!`. This is mostly a
+placeholder RFC since many of the issues affecting the new macro system are
+(or will be) addressed in other RFCs. This RFC may be expanded at a later date.
+
+Currently in this RFC:
+
+* That we should have a new macro by example system,
+* a new keyword for declaring macros.
+
+In other RFCs:
+
+* Naming and modularisation (#1561).
+
+May be added to this RFC later (or might be separate RFCs):
+
+* more detailed syntax proposal,
+* hygiene improvements.
+
+Note this RFC does not involve procedural macros (aka syntax extensions).
+
+
+# Motivation
+[motivation]: #motivation
+
+There are several changes to the macro by example system which are desirable but
+backwards compatible (See [RFC 1561](https://github.com/rust-lang/rfcs/pull/1561)
+for some changes to macro naming and modularisation, I would also like to
+propose improvements to hygiene in macros, and some improved syntax).
+
+In order to maintain Rust's backwards compatibility guarantees, we cannot change
+the existing system (`macro_rules!`) to accommodate these changes. I therefore
+propose a new macro by example system to live alongside `macro_rules!`.
+
+Example (possible) improvements:
+
+```rust
+// Naming (RFC 1561)
+
+fn main() {
+    a::foo!(...);
+}
+
+mod a {
+    // Macro privacy (TBA)
+    pub macro! foo { ... }
```

Comment:
@joshtriplett Yeah, it might be necessary for name clashes. @eternaleye also brought up the fact that with procedural macros you would usually want to import the macro, but you might also want to import the fn that implements the macro, which has the same name, so you could call it as a subroutine of another procedural macro.